### PR TITLE
New formatted log output

### PR DESCRIPTION
**Pull request by aleksei-burlakov**
This is the revival of https://github.com/ClusterLabs/pacemaker/pull/1849

The formatter doesn't differ much, but this PR doesn't introduce new message handler, it uses instead the existing for text ones.

### CODE REVIEW

File: lib/common/output_log.c [hunk 1]
```diff
@@ -0,0 +1,357 @@
...
+
+    priv = (private_data_t *)out->priv;
+    /* backup the log_level*/
+    log_level = priv->attr.log_level;
+    priv->attr.log_level = LOG_ERR;
+    /* print out the log message with the LOG_ERR severity */
+    pcmk__output_crm_log(out, "%s", buffer);
+    /* restore the log_level backup */
+    priv->attr.log_level = log_level;
+    free(buffer);
```
Comment (kgaillot) [inline lib/common/output_log.c]: This function can keep using LOG_ERR for obvious reasons, but it should be sufficient to just call crm_err() directly

File: lib/common/output_log.c [hunk 2]
```diff
@@ -0,0 +1,357 @@
...
+static void
+log_output_message(private_log_message_t *message) {
+    qb_log_from_external_source(message->attr.function
+                                , message->attr.filename
+                                , "%s"
+                                , message->attr.log_level
+                                , message->attr.lineno
+                                , 0
+                                , message->text);
+}
```
[the remark this answers is not in the document]
Comment (aleksei-burlakov) [reply lib/common/output_log.c]: Done.

File: lib/common/output_log.c [hunk 3]
```diff
@@ -0,0 +1,357 @@
...
+    strcpy(message->attr.function, priv->attr.function);
+    message->attr.lineno = priv->attr.lineno;
+
+    va_start(ap, format);
+    len = vasprintf(&message->text, format, ap);
+    CRM_ASSERT(len >= 0);
+    va_end(ap);
+
+    /* save the message for future use */
+    g_queue_push_tail(priv->messages, message);
```
[the remark this answers is not in the document]
Comment (aleksei-burlakov) [reply lib/common/output_log.c]: Done.

File: lib/common/output_log.c [hunk 4]
```diff
@@ -0,0 +1,357 @@
...
+    CRM_ASSERT(out && out->priv);
+    if (proc_stdout != NULL) {
+        message.text = (char *)proc_stdout;
+        log_output_message(&message);
+    }
+
+    if (proc_stderr != NULL) {
+        message.text = (char *)proc_stderr;
+        log_output_message(&message);
+    }
```
[the remark this answers is not in the document]
Comment (aleksei-burlakov) [reply lib/common/output_log.c]: Done.

File: lib/common/output_log.c [hunk 5]
```diff
@@ -0,0 +1,357 @@
...
+
+static void
+log_version(pcmk__output_t *out, bool extended) {
+    /* Make sure the out->attr is already set */
+    if (extended) {
+        pcmk__output_crm_log(out, "Pacemaker %s (Build: %s): %s\n",
+                   PACEMAKER_VERSION, BUILD_VERSION, CRM_FEATURES);
+    } else {
+        pcmk__output_crm_log(out, "Pacemaker %s\n", PACEMAKER_VERSION);
+        pcmk__output_crm_log(out, "Written by Andrew Beekhof\n");
```
[the remark this answers is not in the document]
Comment (aleksei-burlakov) [reply lib/common/output_log.c]: Fixed.

File: lib/common/output_log.c [hunk 6]
```diff
@@ -0,0 +1,357 @@
...
+
+    va_start(ap, format);
+    len = vasprintf(&buffer, format, ap);
+    CRM_ASSERT(len >= 0);
+    va_end(ap);
+
+    if(strcmp(buffer, "") != 0) { /* We don't want empty prefixes */
+        priv->prefixes = g_list_append(priv->prefixes, crm_strdup_printf("%s", buffer));
+    }
+    free(buffer);
```
[the remark this answers is not in the document]
Comment (aleksei-burlakov) [reply lib/common/output_log.c]: Fixed.

File: lib/common/output_log.c [hunk 7]
```diff
@@ -0,0 +1,357 @@
...
+            if (strcmp(prefix, "") != 0) {
+                pcmk__output_crm_log(out, "%s: %s: %s: %s", priority, prefix, name, buffer);
+            } else {
+                pcmk__output_crm_log(out, "%s: %s: %s", priority, name, buffer);
+            }
+        } else {
+            if (strcmp(prefix, "") != 0) {
+                pcmk__output_crm_log(out, "%s: %s: %s", priority, prefix, buffer);
+            } else {
+                pcmk__output_crm_log(out, "%s: %s", priority, buffer);
```
[the remark this answers is not in the document]
Comment (aleksei-burlakov) [reply lib/common/output_log.c]: Fixed.

File: lib/common/output_log.c [hunk 8]
```diff
@@ -0,0 +1,357 @@
...
+    GList *last_prefix = NULL;
+    CRM_ASSERT(priv != NULL);
+    if (priv->prefixes == NULL) {
+      return;
+    }
+    last_prefix = g_list_last(priv->prefixes);
+    CRM_ASSERT(last_prefix != NULL);
+
+    free((char *)last_prefix->data);
+    priv->prefixes = g_list_remove(priv->prefixes, last_prefix->data);
```
[the remark this answers is not in the document]
Comment (aleksei-burlakov) [reply lib/common/output_log.c]: Done.

File: tools/crm_mon_output.c [hunk 9]
```diff
@@ -1154,22 +1164,27 @@ static pcmk__message_entry_t fmt_functions[] = {
     { "cluster-counts", "console", cluster_counts_text },
     { "cluster-counts", "html", cluster_counts_html },
     { "cluster-counts", "text", cluster_counts_text },
+    { "cluster-counts", "log", cluster_counts_text },
```
[the remark this answers is not in the document]
Comment (aleksei-burlakov) [reply tools/crm_mon_output.c]: Done. I indeed used those message handlers for testing the formatter and was not sure if they are required.

File: lib/pengine/pe_output.c [hunk 10]
```diff
@@ -101,22 +101,38 @@ pe__ticket_xml(pcmk__output_t *out, va_list args) {
...
+    uint32_t lineno = va_arg(args, uint32_t);
+    pcmk__output_set_file_func_line(out, filename, function, lineno);
+    return 0;
+}
+
 static pcmk__message_entry_t fmt_functions[] = {
     { "bundle", "xml",  pe__bundle_xml },
     { "bundle", "html",  pe__bundle_html },
     { "bundle", "text",  pe__bundle_text },
+    { "bundle", "log",  pe__bundle_text },
```
[the remark this answers is not in the document]
Comment (aleksei-burlakov) [reply lib/pengine/pe_output.c]: thanks:)

File: lib/common/output_log.c (hunk 1 above)
Comment (aleksei-burlakov) [reply lib/common/output_log.c]: Done. The crm_err is used instead.

**Comment by aleksei-burlakov** [comment]
@kgaillot , I have amended the PR. Could you please make another round of review?

File: lib/common/output_log.c [hunk 11]
```diff
@@ -0,0 +1,267 @@
...
+    int len = 0;
+    va_list ap;
+    char *buffer = NULL;
+
+    va_start(ap, format);
+    len = vasprintf(&buffer, format, ap);
+    CRM_ASSERT(len >= 0);
+    va_end(ap);
+
+    qb_log_from_external_source(NULL, NULL, "%s", LOG_INFO, 0, 0, buffer);
```
Comment (kgaillot) [inline lib/common/output_log.c]: Seeing it reduced, I realize we don't even need pcmk__output_crm_log(), we can just use crm_info() directly

File: lib/common/output_log.c [hunk 12]
```diff
@@ -0,0 +1,267 @@
...
+    int len = 0;
+    va_list ap;
+    char* buffer = NULL;
+
+    va_start(ap, format);
+    len = vasprintf(&buffer, format, ap);
+    CRM_ASSERT(len >= 0);
+    va_end(ap);
+
+    log_list_item(out, NULL, "%s", buffer);
```
Comment (kgaillot) [inline lib/common/output_log.c]: Why use a list item? Is there a reason to use list prefixes in this case?

File: lib/common/output_log.c (hunk 11 above)
Comment (aleksei-burlakov) [reply lib/common/output_log.c]: Done.

File: lib/common/output_log.c (hunk 12 above)
Comment (aleksei-burlakov) [reply lib/common/output_log.c]: Fixed. No, there is no actually.

**Comment by aleksei-burlakov** [comment]
> This looks great, I like the simplified form. To see it in use, you'll have to modify pcmk__schedule_actions() to use it. You can call pcmk__output_new() with "log" as a literal, and replace the print() call with a message call like at the very end of pe_output.c (no options needed).

@kgaillot, should it be in this PR, or in another one?

I have added https://github.com/ClusterLabs/pacemaker/pull/1945/commits/131b5a92f202465237968c7ac75e568100949296 that makes the crm_simulate to use the message functions instead of the sc->fns->print(). There, the begin_list and end_list methods are never called, only the list_item.

File: tools/crm_mon.c [hunk 13]
```diff
@@ -87,6 +87,7 @@ static pcmk__supported_format_t formats[] = {
     CRM_MON_SUPPORTED_FORMAT_CURSES,
 #endif
     PCMK__SUPPORTED_FORMAT_HTML,
+    PCMK__SUPPORTED_FORMAT_LOG,
```
Comment (kgaillot) [inline tools/crm_mon.c]: Drop this line, we won't expose the log formatter to end users

Comment (aleksei-burlakov) [reply tools/crm_mon.c]: Done.

> This is perfect, with one trivial tweak. I tested it on a few of the scheduler test inputs and I think it looks great.

Thanks)

**Comment by kgaillot** [comment]
OK, I think we're almost to where you wanted to be when all this started :)

@clumens will soon move most of what's now in crm_mon_output.c into libpe_status. That's the first step to getting the guts of crm_mon callable from libpacemaker. Once he's done with that we can figure out what the libpacemaker interface should look like.

**Pull request merged**